### PR TITLE
Update Hydra JsonSchema context property possible types

### LINKED ISSUE #4222: JsonSchema assertion will always fail for API Resources that have an output class defined
**API Platform version(s) affected**: 2.5.8 and up

**Description**  

The [SchemaFactory](https://github.com/api-platform/core/blob/main/src/Hydra/JsonSchema/SchemaFactory.php) for Hydra properties is hardcoded to add `@context` as a `string` type. This works fine when the Resource class matches the output object. The [ItemNormalizer](https://github.com/api-platform/core/blob/main/src/JsonLd/Serializer/ItemNormalizer.php) will set `@context` as an IRI/URI in this case.

In case of a Resource class using an output class (see example below) the `ItemNormalizer` will ignore the object and the [ObjectNormalizer](https://github.com/api-platform/core/blob/main/src/JsonLd/Serializer/ObjectNormalizer.php) takes over and will hard-code set the `@context` as an `array`.  This happens via `getAnonymousResourceContext` used in [ContextBuilder](https://github.com/api-platform/core/blob/main/src/JsonLd/ContextBuilder.php#L129). 

**How to reproduce**  

1. Create a resource with output option:

```yaml
Domain\Licence\Entity\LicenceCommentary:
    description: Licence commentary
    attributes:
        output: Domain\Licence\Model\LicenceCommentaryOutput
        input: Domain\Licence\Model\LicenceCommentaryInput
```

2. use the JsonSchema assertions provided by API Platform: 

```php
self::assertMatchesResourceItemJsonSchema(LicenceCommentary::class);
```
```
✘ Create [405 ms]
   │
   │ Failed asserting that Array &0 (
   │     '@context' => Array &1 (
   │         '@vocab' => 'http://example.com/docs.jsonld#'
   │         'hydra' => 'http://www.w3.org/ns/hydra/core#'
   │         'uuid' => 'LicenceCommentaryOutput/uuid'
   │         'type' => 'LicenceCommentaryOutput/type'
   │         'text' => 'LicenceCommentaryOutput/text'
   │     )
   │     '@type' => 'LicenceCommentary'
   │     '@id' => '/licences/57e4525d-3bc7-44ea-a5d9-8039f8f431fa/commentaries/92fd7512-70ff-423c-9136-96529a27b21e'
   │     'uuid' => '92fd7512-70ff-423c-9136-96529a27b21e'
   │     'type' => 'create type'
   │     'text' => 'create text'
   │ ) matches the provided JSON Schema.
   │ @context: Object value found, but a string is required
   │ 
   │ /builds/audi/vms-api/vendor/api-platform/core/src/Bridge/Symfony/Bundle/Test/ApiTestAssertionsTrait.php:109
   │ /builds/audi/vms-api/vendor/api-platform/core/src/Bridge/Symfony/Bundle/Test/ApiTestAssertionsTrait.php:123
   │ /builds/audi/vms-api/tests/Functional/Licence/LicenceCommentaryResourceTest.php:37
   │ 
```

**Possible Solution**  

I have already tried a solution locally, it seems if the Hydra JsonSchema, SchemaFactory becomes smarter by using the ResourceClassInfoTrait it can resolve wether `@context` will be an `object` or `string` via `isResourceClass(string $class)` in the trait. The assertions pass afterwards.

I can open a PR if this is enough.

### PR DESCRIPTION
**Pull request by llupa**
| Q             | A
| ------------- | ---
| Branch?       |  2.6 
| Bug fix?      | yes
| New feature?  |no 
| Deprecations? | no
| Tickets       | Fix #4222
| License       | MIT

JsonSchema assertion will always fail for API Resources that have an output class defined.

### CODE REVIEW

**Comment by llupa** [comment]
I got lost in the test classes! Can someone show me where is the best place to add a test for this?

File: src/Hydra/JsonSchema/SchemaFactory.php [hunk 1]
```diff
@@ -35,7 +35,13 @@ final class SchemaFactory implements SchemaFactoryInterface
         '@type' => self::BASE_PROP,
     ];
     private const BASE_ROOT_PROPS = [
-        '@context' => self::BASE_PROP,
+        '@context' => [
+            'readOnly' => true,
+            'oneOf' => [
+                ['type' => 'string'],
+                ['type' => 'object'],
```
Comment (alanpoulain) [inline src/Hydra/JsonSchema/SchemaFactory.php:42]: Could we have a more detailed object, with required and optional properties?

**Comment by alanpoulain** [comment]
`SchemaFactoryTest` is a good start for the tests.

**Comment by llupa** [comment]
@alanpoulain Hello. I updated the PR. I saw that this: [ContextBuilderTest](https://github.com/api-platform/core/blob/main/tests/JsonLd/ContextBuilderTest.php#L154) is checking that the body of an anon. class normalization has an 'object' `@context`. 

I was thinking to add an integration test where I assert the schema for both a resource class and an anon. class, but I am not sure how to use the fake Models in `tests` 🤔

**Comment by judmir** [comment]
We have the exact same issue at our project. Would be really great if this PR can be merged.

**Comment by alanpoulain** [comment]
I've added an integration test in `ApiTestCaseTest` (I think it's the best place to do so).

**Review by alanpoulain — APPROVED** [review]

**Comment by alanpoulain** [comment]
Thank you @llupa.

**Pull request merged**